lib/class-wp-rest-autosaves-controller.php

```diff
@@ -188,6 +188,10 @@ public function create_item( $request ) {
 		$user_id           = get_current_user_id();
 
 		if ( ( 'draft' === $post->post_status || 'auto-draft' === $post->post_status ) && $post->post_author == $user_id ) {
+			// To avoid resetting the draft's parent, clear the parent of the
+			// prepared revision which is assigned the draft's own ID.
+			unset( $prepared_post->post_parent );
```

Comment:
Is this truly necessary, or are you simply adding it to guard against a client causing a similar bug in the future?

The reason I ask is more from the perspective of: why is `post_parent` a special case? Is it a special case simply because Gutenberg sent it, or is there some additional functional value?

If the former, then we'd need to guard against _all_ extraneous post fields. Which means it probably only makes sense to whitelist the fields we do allow: `title`, `excerpt`, `content`.

Let me know if this doesn't make sense or I missed something.

Reply:
I don't see it as something specific to Gutenberg: In sending a payload to the autosave endpoint, I consider that the `parent_id` value I send should correspond to the post for which we're creating an autodraft revision. That the REST API interprets the fact that it's a draft and thus doesn't need to operate on a revision should change its treatment of the parent field. It's a bit of an overloaded use of the parent field.

>we'd need to guard against all extraneous post fields. Which means it probably only makes sense to whitelist the fields we do allow: title, excerpt, content.

I agree. This would also help avoid setting misleading expectations about the ability to send other fields for this specific draft condition, which won't work for when creating/updating a revision†.

It still needs to accept `parent_id`, but it only becomes relevant when creating a revision in knowing where to attach the revision, not for the update flow here.

† _I've not actually confirmed whether this behavior takes place_

Reply:
@aduth As it turns out, it looks like you flagged this issue 5 months ago: https://core.trac.wordpress.org/ticket/43316#comment:89

---

@adamsilverstein Do you recall why you introduced this conditional logic?

```php
if ( ( 'draft' === $post->post_status || 'auto-draft' === $post->post_status ) && $post->post_author == $user_id ) {
	// Draft posts for the same author: autosaving updates the post and does not create a revision.
	// Convert the post object to an array and add slashes, wp_update_post expects escaped array.
	$autosave_id = wp_update_post( wp_slash( (array) $prepared_post ), true );
} else {
	// Non-draft posts: create or update the post autosave.
	$autosave_id = $this->create_post_autosave( (array) $prepared_post );
}
```

Ideally, we'd prefer to use `$this->create_post_autosave()` for both instances. Barring that, we'll need to replicate the behavior of `_wp_post_revision_data()` prior to calling `wp_update_post()`.